Prepare the `bootstrap` tool for the new check-cfg syntax

Urgau opened the pull request:
This PR prepare the `bootstrap` tool for the [new check-cfg syntax](https://github.com/rust-lang/rust/pull/111072) as well as the according [changes to Cargo](https://github.com/rust-lang/cargo/pull/12845).

~~Note that while the new syntax can technically available on stage > 2, we actually cannot use it since we need a cargo version that supports the new syntax which won't happen until the next beta bump (if I understand everything correctly).~~

r? bootstrap

Urgau committed:
Prepare bootstrap tool for new check-cfg syntax

weihanglo commented:
Hit this issue <https://github.com/rust-lang/rust/pull/117000#issuecomment-1773657187>. How can we help this moving forward?

Urgau commented:
> Hit this issue [#117000 (comment)](https://github.com/rust-lang/rust/pull/117000#issuecomment-1773657187). How can we help this moving forward?

This PR shouldn't be necessary until the next beta bump. Something unexpected is going on.

Urgau committed:
Use local_rebuild property to support new check-cfg in case of rebuild

Kobzol commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

Urgau committed:
Partial Revert: "Auto merge of #117018 - Kobzol:opt-dist-cargo-stage0"

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6613793442/job/17965646880)
Build commit: b8942cea92b3e99ee039753464783527b3fc3722 (`b8942cea92b3e99ee039753464783527b3fc3722`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6613793442/job/17965646880"},"merge_sha":"b8942cea92b3e99ee039753464783527b3fc3722"} -->

rust-timer commented:
Queued b8942cea92b3e99ee039753464783527b3fc3722 with parent a56bd2b944aed6b4a0507cc1870fb4c86d08e48e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722).
There are currently 3 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~4.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

Urgau commented:
I've added a commit to enable the new `{-Z,--}check-cfg` arguments when doing a "local rebuild" (build the same version of rustc with it-self) and also added a revert commit of https://github.com/rust-lang/rust/pull/117018 since it's no longer necessary.

rust-timer commented:
Finished benchmarking commit (b8942cea92b3e99ee039753464783527b3fc3722): [comparison URL](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.6% | [3.6%, 3.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.6% | [3.6%, 3.6%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.1%, 3.0%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.5% | [3.5%, 3.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.5% | [3.5%, 3.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a56bd2b944aed6b4a0507cc1870fb4c86d08e48e&end=b8942cea92b3e99ee039753464783527b3fc3722&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 304.30 MiB -> 304.32 MiB (0.01%)

Urgau commented:
r? @Kobzol (or @Mark-Simulacrum given you suggested the 2nd commit)

Kobzol commented on src/bootstrap/src/bin/rustdoc.rs:
```diff
@@ -70,7 +70,9 @@ fn main() {
     }
     cmd.arg("-Zunstable-options");
+    // #[cfg(bootstrap)]
     cmd.arg("--check-cfg=values(bootstrap)");
```
I wonder if this should be also gated for local rebuilds, but I don't think we can even set a custom rustdoc in `config.toml`, so it probably doesn't matter.

Kobzol reviewed (commented):
I tried a few configurations locally and they were all working.

Hopefully, this will fix bootstrap timings on perf.RLO, so we should do that as soon as possible. Let's merge it and see if anything else breaks.

Kobzol commented:
@bors r+

bors commented:
:pushpin: Commit 0691c06d9401649644dc26acc9285a156ff8d12d has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 0691c06d9401649644dc26acc9285a156ff8d12d -->
<!-- homu: {"type":"Approved","sha":"0691c06d9401649644dc26acc9285a156ff8d12d","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0691c06d9401649644dc26acc9285a156ff8d12d with merge 056f5b0f138fa9556027251fb27e8929350d6191...
<!-- homu: {"type":"BuildStarted","head_sha":"0691c06d9401649644dc26acc9285a156ff8d12d","merge_sha":"056f5b0f138fa9556027251fb27e8929350d6191"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6651755344/job/18078364514)
Approved by: Kobzol
Pushing 056f5b0f138fa9556027251fb27e8929350d6191 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6651755344/job/18078364514"},"merge_sha":"056f5b0f138fa9556027251fb27e8929350d6191"} -->

lqd commented:
![image](https://github.com/rust-lang/rust/assets/247183/76ee2274-8bc9-4b3a-96f6-65345691e2cd)

🎉
